src/components/actions/extendApplicationRoute.ts

```diff
@@ -0,0 +1,63 @@
+import { combineRequestParams } from '../utilityFunctions'
+import { PermissionPolicyType, Trigger } from '../../generated/graphql'
+import { DateTime } from 'luxon'
+import DBConnect from '../databaseConnect'
+import { getPermissionNamesFromJWT } from '../data_display/helpers'
+
+export const routeExtendApplication = async (request: any, reply: any) => {
+  const { applicationId, eventCode, extensionTime, data } = combineRequestParams(request, 'camel')
+
+  // Check permissions first -- currently we just check user has ANY Reviewing
+  // permissions for current template.
+  // TO-DO: define specific permission names required for extending deadlines
+  const { permissionNames: userPermissions } = await getPermissionNamesFromJWT(request)
+
+  const templatePermissions = (
+    await DBConnect.getTemplatePermissionsFromApplication(Number(applicationId))
+  )
+    .filter(
+      (permission: any) =>
+        permission?.permissionName?.permissionPolicy?.type === PermissionPolicyType.Review
+    )
+    .map((permission: any) => permission?.permissionName?.name)
+
+  const hasPermission = userPermissions.some((permission) =>
+    templatePermissions.includes(permission)
+  )
+
+  if (!hasPermission) return reply.send({ success: false, message: 'Unauthorized' })
+
+  try {
+    // If time is a number, we consider it as number of days. Otherwise a Luxon
+    // duration object can be provided for more specificity
+    const scheduledTime = isNaN(Number(extensionTime))
+      ? DateTime.now().plus(extensionTime).toISO()
+      : DateTime.now()
+          .plus({ days: Number(extensionTime) })
+          .toISO()
+
+    // Update trigger schedule event, return error if event doesn't exist
+    const extensionResult = await DBConnect.updateScheduledEventTime(
+      applicationId,
+      eventCode,
+      scheduledTime
+    )
+
+    if (extensionResult.length === 0)
+      return reply.send({ success: false, message: 'No matching event found' })
+
+    // Add trigger event directly to trigger_queue so we can include eventCode
```

Comment:
Could have gone one of three ways here:
- "simulate" a trigger by sending directly to "processTrigger" the same as the "preview" endpoint does
- set a real trigger on the "application" table and have it act as normal
- set a trigger record directly, and then the database listeners proceed as normal

Went with the last option, as it was the easiest way to get the eventCode (and other possible data) passed through while keeping action behaviour as "standard" as possible

Reply:
hmm i noticed that in front end we do `reload`, which usually waits for application trigger to be reset. When we add trigger directly, do we set trigger on application ? (is it possible that ui won't update with correct 'new' application deadline if processing triggers take a bit longer ?)

Reply:
If I remember correctly, the front-end doesn't call "reload" until after it receives a valid response from this endpoint. And this endpoint doesn't send response until everything is completed.

Reply:
ok, so small chance that it will still show expired, maybe it's good to make an issue to take note of that. I would assume we can add another trigger_queue_status 'completed' and await for that (or error) before sending response to api call

Reply:
> I would assume we can add another trigger_queue_status 'completed' and await for that (or error) before sending response to api call

Actually, I guess it wouldn't be that hard to just add a bit extra here to:
- wait for "trigger_queue" event status to be "ACTIONS_DISPATCHED"
- wait for the actions added to the "action_queue" that match this trigger_queue id to become no longer "PROCESSING"

I'll do that now, shouldn't be too tricky.

Reply:
Okay @andreievg -- here you go:
089dc9d6

Made a generic "waitForDatabaseValue" method which can be re-used for any other "waiting" events we need.